octoprint_firmwareupdate/__init__.py

```diff
@@ -1,85 +1,271 @@
 # coding=utf-8
 from __future__ import absolute_import
 
-__author__ = "Kevin Murphy <kevin@voxel8.co>"
-__license__ = "GNU Affero General Public License http://www.gnu.org/licenses/agpl.html"
-__copyright__ = "Copyright (C) 2015 Kevin Murphy - Released under terms of the AGPLv3 License"
-
 import octoprint.plugin
-from octoprint.util import RepeatedTimer
 import flask
 import os
-from subprocess import Popen, call
-import signal
-import linecache
+from subprocess import Popen
 import psutil
-import time
+from time import sleep
+import requests
+import urllib
+from threading import Thread
+from glob import glob
+from serial import Serial, SerialException
+from octoprint.events import eventManager, Events
+
+Events.FIRMWARE_UPDATE = "FirmwareUpdate"
+
+__author__ = "Kevin Murphy <kevin@voxel8.co>"
+__license__ = "GNU Affero General Public License http://www.gnu.org/licenses/agpl.html"
+__copyright__ = "Copyright (C) 2015 Kevin Murphy - Released under terms of the AGPLv3 License"
+
 
 class FirmwareUpdatePlugin(octoprint.plugin.StartupPlugin,
-                       octoprint.plugin.TemplatePlugin,
-                       octoprint.plugin.AssetPlugin,
-                       octoprint.plugin.SettingsPlugin,
-                       octoprint.plugin.SimpleApiPlugin):
+                           octoprint.plugin.TemplatePlugin,
+                           octoprint.plugin.AssetPlugin,
+                           octoprint.plugin.SettingsPlugin,
+                           octoprint.plugin.SimpleApiPlugin):
 
     def __init__(self):
         self.isUpdating = False
+        self.firmware_file = None
+        self.version_file = None
         self._checkTimer = None
         self.updatePID = None
+        self.f = None
+        self.completion_time = 0
+        self.write_time = None
+        self.read_time = None
+        self.port = None
+        self.version = None
 
     def get_assets(self):
         return {
             "js": ["js/firmwareupdate.js"]
         }
 
-    def startTimer(self, interval):
-        self._checkTimer = RepeatedTimer(interval, self.checkStatus, run_first=True, condition=self.checkStatus)
-        self._checkTimer.start()
+    def get_api_commands(self):
+        return dict(
+            update_firmware=[]
+        )
+
+    def on_api_command(self, command, data):
+        if command == "update_firmware":
+            self._start_update()
+
+        else:
+            self._logger.info("Uknown command.")
```

Comment:
Typo: Uknown -> Unknown

Also, it might be a good idea to put the unknown command in the log msg.